opensrp-app/src/main/java/org/smartregister/service/HTTPAgent.java

```diff
@@ -189,48 +197,62 @@ public void setupHttpClient() {
 
     public LoginResponse urlCanBeAccessWithGivenCredentials(String requestURL, String userName,
                                                             String password) {
-        HttpResponse httpResponse = null;
         LoginResponse loginResponse = null;
-
+        HttpURLConnection urlConnection = null;
         try {
-            setCredentials(userName, password);
             requestURL = requestURL.replaceAll("\\s+", "");
-            httpResponse = httpClient.execute(new HttpGet(requestURL));
-            int statusCode = httpResponse.getStatusLine().getStatusCode();
+            URL url = new URL(requestURL);
+            urlConnection = (HttpURLConnection) url.openConnection();
+            if (urlConnection instanceof HttpsURLConnection) {
+                OpensrpSSLHelper opensrpSSLHelper = new OpensrpSSLHelper(context, configuration);
+                ((HttpsURLConnection) urlConnection).setSSLSocketFactory(opensrpSSLHelper.getSSLSocketFactory());
+            }
+            urlConnection.setConnectTimeout(CONNECTION_TIMEOUT);
+            urlConnection.setReadTimeout(READ_TIMEOUT);
+            final String basicAuth = "Basic " + Base64.encodeToString((userName + ":" + password).getBytes(), Base64.NO_WRAP);
```

Comment:
Do you think this would be cleaner you handled it using [UsernamePasswordCredentials](https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/auth/UsernamePasswordCredentials.html)?

Reply:
This class is for deplicated  apache http client library.

The alternative is to use` java.net.PasswordAuthentication` with `java.net.HttpURLConnection` but this has a bug on the underlying okhttp library where the credentials are invalid.

It keeps on retrying to authenticate in case the the server does not authorize a request.

Ones gets the below error

`java.net.ProtocolException: Too many follow-up requests: 21`

Reply:
OK, cool.